### ISSUE TITLE
Is it possible to improve cv::cuda::remap performance by using Texture Mapping Units?

### ISSUE THREAD

**Issue by aftersomemath** [opencv/opencv#23312]
### Descripe the feature and motivation

Graphics cards contain texture mapping units which are dedicated hardware for [remapping and interpolating textures](https://en.wikipedia.org/wiki/Texture_mapping_unit).

It can be seen in [cuda/filters.hpp](https://github.com/opencv/opencv/blob/0052d46b8e33c7bfe0e1450e4bff28b88f455570/modules/core/include/opencv2/core/cuda/filters.hpp) which is used by [cuda/remap.cu](https://github.com/opencv/opencv_contrib/blob/ed1873bc2c58f1c2dc94f98c816be0d39068995f/modules/cudawarping/src/cuda/remap.cu) that OpenCV's cuda remapping done in software without taking advantage of the texture mapping unit.

It seems likely that the performance of `cv::cuda::remap` would be could be greatly improved if it used the dedicated hardware designed for texture remapping. Is there any technical reason why this can not be done?

CUDA appears to expose access to the texture mapping unit. [This presentation](https://on-demand.gputechconf.com/gtc-express/2011/presentations/texture_webinar_aug_2011.pdf) is a little old but explains the advantages well.

**Comment by cudawarped** [comment]
@aftersomemath Texture memory is used when no `cudaStream_t`  is passed for one (uchar,short,ushort,float) and four (uchar4,short4,ushort4,float4) channel data.  You may have missed that the places where the filter is initialized with a texture object e.g. [here](https://github.com/opencv/opencv_contrib/blob/ed1873bc2c58f1c2dc94f98c816be0d39068995f)
```
cudev::Texture<T> texSrcWhole(srcWhole);
Filter<cudev::TexturePtr<T>> filter_src(texSrcWhole);
```

I am not sure but it would be interesting to find out if texture memory is still any faster for this and other operations on modern hardware.  A first step would be to compare the performance of the texture version to the non texture version, say float4 vs float3.  If the performance on the more computationally expensive datatype (float4) in texture memory is the same or better that could imply there is some advantage.  Unfortunately as this is implementation dependant so futher testing would be needed but it would be a good indicator.

**Comment by aftersomemath** [comment]
Great point! I was wondering about `RemapDispatcherNonStreamTex`.

I see that the filter is initialized with a texture object which presumably can use the texture mapping unit (TMU) based on what's going on in [cudev/ptr2d/texture.hpp](https://github.com/opencv/opencv_contrib/blob/ed1873bc2c58f1c2dc94f98c816be0d39068995f/modules/cudev/include/opencv2/cudev/ptr2d/texture.hpp#L31) and the docs for [cudaTextureObject_t](https://docs.nvidia.com/cuda/cuda-runtime-api/group__CUDART__TEXTURE__OBJECT.html).

However, if I'm not mistaken, these `cudev::TexturePtr`'s are accessed through a `PointFilter`, `LinearFilter`, or `CubicFilter` as implemented in [cuda/filters.hpp](https://github.com/opencv/opencv/blob/0052d46b8e33c7bfe0e1450e4bff28b88f455570/modules/core/include/opencv2/core/cuda/filters.hpp). So, it seems the TMU's interpolation capabilities still aren't used when texture memory is used to back the array in `cuda/remap.cu`.

Also `PointFilter` contains a [__float2int](https://github.com/opencv/opencv/blob/0052d46b8e33c7bfe0e1450e4bff28b88f455570/modules/core/include/opencv2/core/cuda/filters.hpp#L74) call that stops it from being used as a "passthrough" to get the TMU to do the work.

If all of this is correct (I'm not super confident in myself here), then a simple test would be to hack `remap.cu` until the top level remap kernel can accept a `cudev::TexturePtr` directly and then run some benchmarks.

Does all of this sound right?

**Comment by cudawarped** [comment]
> However, if I'm not mistaken, these `cudev::TexturePtr`'s are accessed through a `PointFilter`, `LinearFilter`, or `CubicFilter` as implemented in [cuda/filters.hpp](https://github.com/opencv/opencv/blob/0052d46b8e33c7bfe0e1450e4bff28b88f455570/modules/core/include/opencv2/core/cuda/filters.hpp). So, it seems the TMU's interpolation capabilities still aren't used when texture memory is used to back the array in `cuda/remap.cu`.

I'm not sure I follow.  The `cudev::TexturePtr` is stored in [src](https://github.com/opencv/opencv/blob/0052d46b8e33c7bfe0e1450e4bff28b88f455570/modules/core/include/opencv2/core/cuda/filters.hpp#L86) and then read from inside the overloaded [operator ()](https://github.com/opencv/opencv/blob/0052d46b8e33c7bfe0e1450e4bff28b88f455570/modules/core/include/opencv2/core/cuda/filters.hpp#L101) every time memory needs to be read from src?

> Also `PointFilter` contains a [__float2int](https://github.com/opencv/opencv/blob/0052d46b8e33c7bfe0e1450e4bff28b88f455570/modules/core/include/opencv2/core/cuda/filters.hpp#L74) call that stops it from being used as a "passthrough" to get the TMU to do the work.

The memory is still accessed by calls to `tex2D` so it should still be going through the texture unit even if it has been cast to an integer.

> If all of this is correct (I'm not super confident in myself here), then a simple test would be to hack `remap.cu` until the [remap kernel](https://github.com/opencv/opencv_contrib/blob/ed1873bc2c58f1c2dc94f98c816be0d39068995f/modules/cudawarping/src/cuda/remap.cu#L57) can accept a `cudev::TexturePtr` directly and then run some benchmarks.

Again I'm not sure why you would need to do this as remap is already using texture memory.

**Comment by aftersomemath** [comment]
I agree, `cudev::TexturePtr` is stored in `src` and then accessed from the overloaded `()` operator which will end up going through the texture unit.

However, just like `PointFilter`, `LinearFilter` only accesses the `cudev::TexturePtr` at integer coordinates (`x1` and `y1`). Which I think means the interpolation capabilities of the texture unit are being bypassed (because the read is at an integer location).

My understanding is that if `LinearFilter` just returned `src(y, x)` (`y` and `x` being floats) instead of `src(y1, x1)` then the texture unit would be used for interpolation (as long as the texture unit is configured correctly).

Maybe there is some other way that the texture unit is being used to do the interpolation, but since [cuda/filters.hpp](https://github.com/opencv/opencv/blob/0052d46b8e33c7bfe0e1450e4bff28b88f455570/modules/core/include/opencv2/core/cuda/filters.hpp) only accesses the `cudev::TexturePtr` at integer coordinates, I don't see how that's possible.

**Comment by cudawarped** [comment]
> However, just like `PointFilter`, `LinearFilter` only accesses the `cudev::TexturePtr` at integer coordinates (`x1` and `y1`). Which I think means the interpolation capabilities of the texture unit are being bypassed (because the read is at an integer location).
> 
> My understanding is that if `LinearFilter` just returned `src(y, x)` (`y` and `x` being floats) instead of `src(y1, x1)` then the texture unit would be used for interpolation (as long as the texture unit is configured correctly).
> 
> Maybe there is some other way that the texture unit is being used to do the interpolation, but since [cuda/filters.hpp](https://github.com/opencv/opencv/blob/0052d46b8e33c7bfe0e1450e4bff28b88f455570/modules/core/include/opencv2/core/cuda/filters.hpp) only accesses the `cudev::TexturePtr` at integer coordinates, I don't see how that's possible. Am I mistaken?

Now I understand what you are getting at.  Whilst the memory is being read through the texture unit utilizing the texture cache to avoid possible warp coalescing issues it is not using the texure hardware to perform the interpolation.

I am not sure why this is.  Its possible it wouldn't be any faster if the kenel is memory bound but it would be worth checking to make sure.

**Comment by aftersomemath** [comment]
Great! I will try to do a test at some point. I agree it could end up being memory bound for the typical OpenCV use case.
